File: Model/K-means/K-means.r
## !/user/bin/env RStudio 1.1.423
## -*- coding: utf-8 -*-
## K-means Model

library("dplyr")
library('magrittr')
library('ggplot2')

rm(list = ls())
gc()

Data_Input <- function(file_path = "D:/R/File/iris.csv",p = .75){
    data = read.csv(file_path,stringsAsFactors = FALSE,check.names = FALSE)
    names(data) <- c('sepal_length','sepal_width','petal_length','petal_width','class')
    data[,-ncol(data)] <- scale(data[,-ncol(data)])
    x = data[,1:(ncol(data)-2)];y =  data$class_c
    return(
        list(
            data = data,
            train_data = x,
            train_target = y
            )
        )
   }


DistEclud <- function(vecA, vecB){
    diff = rbind(vecA,vecB)
    euclidean = dist(diff) %>% as.numeric()
    return (euclidean)
   }


RandCentre <- function(dataSet, k){  
    n = ncol(dataSet)
    Centres = matrix(nrow = k,ncol = n)
    for(j in 1:n){
        minJ = min(dataSet[,j])
        rangeJ = max(dataSet[,j]) - minJ
        Centres[,j] = (minJ + rangeJ * runif(k)) 
    }
    return (Centres)
} 


Kmeans_Cluster <- function(dataSet,k){
    m = nrow(dataSet)
    ClusterAssment = rep(0,times = 2*m) %>% matrix(nrow = m,ncol = 2)  
    Centres = RandCentre(dataSet,k)
    ClusterChanged = TRUE
    while(ClusterChanged){
        ClusterChanged = FALSE
        for(i in 1:m){
            minDist = Inf
            minIndex = 0
            for(j in 1:k){
                distJI = DistEclud(Centres[j,],dataSet[i,])
                if (distJI < minDist){
                    minDist = distJI
                    minIndex = j
                }
            }
            if(ClusterAssment[i,1] != minIndex){
                  ClusterChanged = TRUE
                  ClusterAssment[i,] = c(minIndex,minDist^2)
            } 
        }
        for(cent in 1:k){
            index = grep(cent,ClusterAssment[,1])
            ptsInClust = dataSet[index,]
            Centres[cent,] = apply(ptsInClust,2,mean)
        }
    print(Centres)
    }
    return (
        list(
            Centres = Centres,
            ClusterAssment = ClusterAssment
            )
        )
}



Show_Result <- function(k){
    data_source = Data_Input()
    dataSet = data_source$train_data
    y = data_source$train_target
    result = Kmeans_Cluster(dataSet,k)
    centroids = result$Centres
    clusterAssment = result$ClusterAssment
    ggplot() +
    geom_point(data = NULL,aes(x = dataSet[,1],y = dataSet[,2],fill = factor(clusterAssment[,1])),shape = 21,colour = 'white',size = 4) +
    geom_point(data = NULL,aes(x= centroids[,1],y = centroids[,2]),fill = 'Red',size = 10,shape = 23) +
    scale_fill_brewer(palette = 'Set1') +
    guides(fill=guide_legend(title=NULL)) +
    theme_void()
   }    
    
Show_Result(3)

File: Model/K-means/K-means.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy  as np
import pandas as pd
from sklearn import preprocessing
from sklearn.metrics import confusion_matrix
from matplotlib import pyplot as plt 

def DataInput():
    data = pd.read_csv("D:/Python/File/iris.csv")
    data.columns = ['sepal_length','sepal_width','petal_length','petal_width','class']
    print(data.shape,'\n',data.head())
    data.iloc[:,0:-1] = preprocessing.scale(data.iloc[:,0:-1])
    data['class_c'] =  pd.factorize(data['class'])[0]
    return  data.iloc[:,0:-2],data.iloc[:,-1]  


def DistEclud(vecA, vecB):
    return np.sqrt(np.sum(np.power(vecA - vecB, 2))) 

def RandCentre(dataSet, k):     
    n = dataSet.shape[1]
    Centres = np.mat(np.zeros((k,n)))
    for j in range(n):
        minJ = np.min(dataSet.iloc[:,j])
        rangeJ = np.float(np.max(dataSet.iloc[:,j]) - minJ)
        Centres[:,j] = np.mat(minJ + rangeJ * np.random.rand(k,1)) 
    return Centres


def kMeans(dataSet,k): 
    m = dataSet.shape[0]
    ClusterAssment = np.mat(np.zeros((m,2)))
    Centres = RandCentre(dataSet,k)
    ClusterChanged = True
    while ClusterChanged:
        ClusterChanged = False
        for i in range(m): 
            minDist = np.inf;minIndex = -1
            for j in range(k):
                distJI = DistEclud(Centres[j,:],dataSet.iloc[i,:].values)
                if distJI < minDist:
                    minDist = distJI
                    minIndex = j
            if ClusterAssment[i,0] != minIndex:
                ClusterChanged = True
                ClusterAssment[i,:] = minIndex,minDist**2
        print(Centres)
        for cent in range(k):
            index = np.nonzero(ClusterAssment[:,0].A== cent)[0].tolist()
            ptsInClust = dataSet.iloc[index,:]
            Centres[cent,:] = np.mean(ptsInClust,axis=0).values
    return Centres,ClusterAssment


def show(k): 
    dataSet,y = DataInput()
    centroids, clusterAssment = kMeans(dataSet,k)
    m,n = dataSet.shape  
    mark  = ['or', 'ob', 'og']
    for i in range(m):  
        markIndex = np.int(clusterAssment[i,0])  
        plt.plot(dataSet.iloc[i,0], dataSet.iloc[i,1],mark[markIndex])  
    for i in range(k):  
        plt.plot(centroids[i,0], centroids[i,1], mark[i], markersize = 12)  
    plt.show()
      
show(k=3)